src/StudyEntities/Candidate/SiteIDGenerator.php

```diff
@@ -0,0 +1,258 @@
+                . "Generation: `{$this->generationMethod}`\n"
+            );
+        }
+    }
+
+    /**
+     * Generates a new ID for use in the rest of LORIS.
+     * This function should be updated to use PSCID and ExternalID classes when
+     * they are created. See CandIDGenerator for an example.
+     *
+     * @return string
+     */
+    public function generate() {
+        return $this->createNewID();
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * The parameters $kind and $prefix should never be user-controllable as
+     * this creates a SQL injection risk.
+     */
+    protected function getExistingIDs(): array
+    {
+        $ids = \Database::singleton()->pselectCol(
+            "SELECT substring($this->kind, LENGTH('{$this->prefix}') +1)
+            from candidate
+            WHERE {$this->kind} LIKE '{$this->prefix}%'",
+            array()
+        );
+
+        return !empty($ids) ? $ids : array();
+    }
+    /**
+     * Helper function used for extracting the values from the config
+     * settings relating to the PSCID structure.
+     *
+     * @param string $setting One of: 'generation', 'length', 'alphabet',
+     * 'length', 'min', 'max'.
+     *
+     * @return array|int|string|null
+     */
+    private function getIDSetting(
+        string $setting
+    ) {
+        // The generation setting can be easily extracted and returned.
+        if ($setting == 'generation') {
+            return \NDB_Factory::singleton()
+                ->config()
+                ->getSetting($this->kind)['generation'];
+        }
+
+        // Values other than 'generation' are found within 'seq' elements and
+        // require more complex processing.
+        $idStructure = \NDB_Factory::singleton()
+            ->config()
+            ->getSetting($this->kind)['structure']['seq'];
+
+        if (!$idStructure[0]) {
+            // There's only one seq tag so the param format
+            // needs to be fixed
+            $temp        = array();
+            $temp[]      = $idStructure;
+            $idStructure = $temp;
+        }
+
+        try {
+            $seqValue = self::getSeqAttribute($idStructure, $setting);
+        } catch (\ConfigurationException $e) {
+            /* Throw a new exception so that we can inform developers whether
+             * the ConfigurationException arose due to ExternalID or PSCID
+             * settings.
+             */
+            throw new \LorisException(
+                "Cannot create new candidate because of a configuration " .
+                "error in settings for {$this->kind} structure. Details: "
+                . $e->getMessage()
+            );
+        }
+        if ($setting === 'alphabet') {
+            switch ($seqValue) {
+            case 'alpha':
+                return range('A', 'Z');
+            case 'numeric':
+                return range(0, 9);
+            case 'alphanumeric':
+                return array_merge(range(0, 9), range('A', 'Z'));
```

Comment:
Is this supposed to be '0' and '9'? This is merging an array of strings with an array of integers, they aren't really comparable types. (Same with numeric above)

Reply:
It doesn't seem to cause issues in my testing but I'll change it to be safe.